src/java/org/apache/cassandra/service/CacheService.java

```diff
@@ -472,7 +482,23 @@ public void serialize(KeyCacheKey key, DataOutputPlus out, ColumnFamilyStore cfs
             int generation = input.readInt();
             input.readBoolean(); // backwards compatibility for "promoted indexes" boolean
             SSTableReader reader = null;
-            if (cfs == null || !cfs.isKeyCacheEnabled() || (reader = findDesc(generation, cfs.getSSTables(SSTableSet.CANONICAL))) == null)
+            if (!skipEntry)
+            {
+                Pair<String, String> qualifiedName = Pair.create(cfs.metadata.ksName, cfs.metadata.cfName);
+                Map<Integer, SSTableReader> generationToSSTableReader = cachedSSTableReaders.get(qualifiedName);
+                if (generationToSSTableReader == null)
+                {
+                    generationToSSTableReader = cfs.collect(SSTableSet.CANONICAL,
```

Comment:
could we just build the map here? Don't see a need for the new collect(...) method beyond this

if we want to keep the method there should probably be tests for it as well